driver-core/src/main/com/mongodb/internal/authentication/SaslPrep.java

```diff
@@ -0,0 +1,292 @@
+     * <a href="https://tools.ietf.org/html/rfc3454#appendix-C.6">RFC 3454, Appendix C.6</a>.
+     */
+    private static boolean inappropriateForPlainText(final int codepoint) {
+        return codepoint == 0xFFF9
+                || codepoint == 0xFFFA
+                || codepoint == 0xFFFB
+                || codepoint == 0xFFFC
+                || codepoint == 0xFFFD;
+    }
+
+    /**
+     * Return true if the given {@code codepoint} is a surrogate code point as defined by
+     * <a href="https://tools.ietf.org/html/rfc3454#appendix-C.5">RFC 3454, Appendix C.5</a>.
+     */
+    private static boolean surrogateCodePoint(final int codepoint) {
+        return 0xD800 <= codepoint && codepoint <= 0xDFFF;
+    }
+
+    /**
+     * Return true if the given {@code codepoint} is a non-character code point as defined by
+     * <a href="https://tools.ietf.org/html/rfc3454#appendix-C.4">RFC 3454, Appendix C.4</a>.
+     */
+    private static boolean nonCharacterCodePoint(final int codepoint) {
+        return 0xFDD0 <= codepoint && codepoint <= 0xFDEF
+                || 0xFFFE <= codepoint && codepoint <= 0xFFFF
+                || 0x1FFFE <= codepoint && codepoint <= 0x1FFFF
+                || 0x2FFFE <= codepoint && codepoint <= 0x2FFFF
+                || 0x3FFFE <= codepoint && codepoint <= 0x3FFFF
+                || 0x4FFFE <= codepoint && codepoint <= 0x4FFFF
+                || 0x5FFFE <= codepoint && codepoint <= 0x5FFFF
+                || 0x6FFFE <= codepoint && codepoint <= 0x6FFFF
+                || 0x7FFFE <= codepoint && codepoint <= 0x7FFFF
+                || 0x8FFFE <= codepoint && codepoint <= 0x8FFFF
+                || 0x9FFFE <= codepoint && codepoint <= 0x9FFFF
+                || 0xAFFFE <= codepoint && codepoint <= 0xAFFFF
+                || 0xBFFFE <= codepoint && codepoint <= 0xBFFFF
+                || 0xCFFFE <= codepoint && codepoint <= 0xCFFFF
+                || 0xDFFFE <= codepoint && codepoint <= 0xDFFFF
+                || 0xEFFFE <= codepoint && codepoint <= 0xEFFFF
+                || 0xFFFFE <= codepoint && codepoint <= 0xFFFFF
+                || 0x10FFFE <= codepoint && codepoint <= 0x10FFFF;
+    }
+
+    /**
+     * Return true if the given {@code codepoint} is a private use character as defined by
+     * <a href="https://tools.ietf.org/html/rfc3454#appendix-C.3">RFC 3454, Appendix C.3</a>.
+     */
+    private static boolean privateUse(final int codepoint) {
+        return 0xE000 <= codepoint && codepoint <= 0xF8FF
+                || 0xF000 <= codepoint && codepoint <= 0xFFFFD
+                || 0x100000 <= codepoint && codepoint <= 0x10FFFD;
+    }
+
+    /**
+     * Return true if the given {@code ch} is a non-ASCII control character as defined by
+     * <a href="https://tools.ietf.org/html/rfc3454#appendix-C.2.2">RFC 3454, Appendix C.2.2</a>.
+     */
+    private static boolean nonAsciiControl(final int codepoint) {
```

Comment:
There's a few transcription errors in this method.  

See https://tools.ietf.org/html/rfc3454.html#appendix-C.2.2